[freshness-policies] Add FreshnessPolicyInfo graphql endpoint

OwenKephart opened the pull request:
### Summary & Motivation

We want to be able to access info about how out of date an asset is with respect to its freshness policy in Dagit. We use a utility function to calculate this value. There are two values that we're interested in in particular (outside of a description of the policy). These are the number of minutes late a given asset currently is, and the number of minutes late that the most recent materialization of the asset was.

We currently can't take advantage of the fancy latestMaterialization batch loader thing, as we need the EventLogRecord (which has a storage id), not just the EventLogEntry. In the short term, this isn't a huge deal as we'll only need to make this query for each asset that has a freshness policy defined (rather than every single asset), but in the future, this should be optimized.

We need to be able to serialize the FreshnessPolicy objects because this will be evaluated in a process that does not have access to the source code, so we need to create these ExternalFreshnessPolicy things (we do a similar thing with PartitionSets). 

### How I Tested These Changes

Hacked in a little bit of typescript and was able to add this info to the asset node live query / asset node in dagit. Worked as expected

clairelin135 commented on python_modules/dagster-graphql/dagster_graphql/schema/asset_graph.py:
```diff
@@ -80,6 +81,15 @@ def resolve_asset(self, _graphene_info):
 
 
+class GrapheneAssetFreshnessInfo(graphene.ObjectType):
+    currentMinutesLate = graphene.Field(graphene.Float)
```
I think these policyDescription field should be required, since it would be populated anytime the `FreshnessInfo` object is returned

so `graphene.Field(graphene.String)` -> `graphene.String`

OwenKephart replied:
good call (looks like I need it to be `graphene.NonNull(graphene.String)`)

clairelin135 replied:
yep, that should do it!

clairelin135 commented on python_modules/dagster-graphql/dagster_graphql/schema/asset_graph.py:
```diff
@@ -529,6 +540,74 @@ def resolve_type(self, _graphene_info) -> Optional[str]:
+        from dagster import DagsterEventType, EventRecordsFilter
+        from dagster._utils.calculate_data_time import get_upstream_materialization_times_for_record
+
+        # no FreshnessPolicy, so freshnessInfo does not apply
```
maybe we can break this out into a helper method in a separate file that returns `GrapheneAssetFreshnessInfo` objects to keep the `GrapheneAssetNode` object short?

clairelin135 reviewed (commented):
This is looking good to me. Can we also add a test for this? (The other asset graphql tests are located in this file: https://github.com/dagster-io/dagster/blob/master/python_modules/dagster-graphql/dagster_graphql_tests/graphql/test_assets.py)

OwenKephart commented on python_modules/dagster/dagster/_core/host_representation/external_data.py:
```diff
@@ -509,6 +515,99 @@ def __new__(cls, error: Optional[SerializableErrorInfo]):
         )
 
 
+class ExternalFreshnessPolicyData(ABC):
```
A thought: how bad would it be to just `@whitelist_for_serdes` each of the actual FreshnessPolicy objects and make them NamedTuples?

sryza replied:
Hmm good question. I can't think of a problem.  Maybe @alangenfeld can?

alangenfeld replied:
yea i don't have full context but i assume these policy objects are declarative little data objects that would be fine as serdes tuples.

sryza commented on python_modules/dagster-graphql/dagster_graphql/implementation/fetch_assets.py:
```diff
@@ -243,3 +254,84 @@ def get_unique_asset_id(
+    asset_graph: AssetGraph,
+    freshness_policy: FreshnessPolicy,
+    batch_loader: BatchMaterializationLoader,
+):
```
A return type would be helpful here

sryza commented on python_modules/dagster/dagster/_core/definitions/asset_graph.py:
```diff
@@ -1,56 +1,122 @@
+    ) -> PartitionMapping:
+        partitions_def = self.get_partitions_def(asset_key)
+        partition_mappings = self.partition_mappings_by_key.get(asset_key) or {}
+        return partition_mappings.get(
```
This is duplicating logic inside `AssetsDefinition.get_partition_mapping`, right? I think worth factoring it out into a common function so we don't accidentally change one without changing the other.

sryza commented on python_modules/dagster/dagster/_core/definitions/asset_graph.py:
```diff
@@ -1,56 +1,122 @@
+            asset_dep_graph={"upstream": upstream, "downstream": downstream},
+            source_asset_keys=set(),  # TODO
+            partitions_defs_by_key=partitions_defs_by_key,
+            partition_mappings_by_key=dict(),  # TODO
```
Can we raise an error if any of the assets have partition mappings or if someone tries to use the partition mapping or something? Otherwise, I think it will be easy to write bugs that expect this to work.

sryza commented on python_modules/dagster-graphql/dagster_graphql/implementation/fetch_assets.py:
```diff
@@ -243,3 +254,84 @@ def get_unique_asset_id(
+        # the latest record
+        original_times = {
+            upstream_key: _convert_timestamp(
+                get_latest_materialization_record(
```
Would we be able to use the `TickInstanceQueryer` (maybe rename it) so we can cache these?

sryza commented on python_modules/dagster-graphql/dagster_graphql/implementation/fetch_assets.py:
```diff
@@ -243,3 +254,84 @@ def get_unique_asset_id(
+    )
+
+    # not all FreshnessPolicies require this information, so don't fetch it if not necessary
+    if isinstance(freshness_policy, MaximumLatencyFreshnessPolicy):
```
Can we include a boolean property on the freshness policy instead of testing for particular subclasses?

clairelin135 reviewed (approved):
Mostly looked at the graphql fields portions here, but those are looking good to me!

OwenKephart commented:
simplified this implementation a bunch based off of https://github.com/dagster-io/dagster/pull/10457, so would be great to get a once-over on this before I go ahead w/ merging it.

also split out GrapheneFreshnessPolicy from GrapheneAssetFreshnessInfo.

clairelin135 commented:
@OwenKephart I think there might be some missing files, I don't see a `GrapheneFreshnessPolicy` object

sryza reviewed (approved):
All the non-GraphQL bits look good to me.

If we were less crunched for time, I'd say the AssetGraph changes should go into the own PR.

clairelin135 reviewed (approved):
GraphQL changes look good to me!
